src/layouts/components/Editor/components/MenuBar.tsx

```diff
@@ -143,19 +276,122 @@ export const MenuBar = ({ editor }: { editor: Editor }) => {
       borderBottom="1px solid"
       borderColor="base.divider.strong"
       borderTopRadius="0.25rem"
+      spacing="0.125rem"
     >
       {items.map((item) => (
         <>
-          {item.type === "divider" ? (
+          {item.type === "divider" && (
             <Divider
               orientation="vertical"
               border="px solid"
               borderColor="base.divider.strong"
               h="1.25rem"
             />
-          ) : (
-            <MenuItem {...item} />
           )}
+
+          {item.type === "vertical-list" && (
+            <Menu>
+              {({ isOpen }) => {
+                const activeItem = item.items.find(
+                  (subItem) => subItem.isActive && subItem.isActive()
+                )
+
+                return (
+                  <>
+                    <Menu.Button
+                      variant="clear"
+                      colorScheme="grey"
+                      isOpen={isOpen}
+                      size="lg"
+                      p="0.75rem"
+                      w={item.buttonWidth}
+                    >
+                      {activeItem?.title || item.defaultTitle}
+                    </Menu.Button>
+
+                    <Menu.List w={item.menuWidth}>
+                      {item.items.map((subItem) => (
+                        <Menu.Item onClick={subItem.action}>
+                          {subItem.leftItem}
+                          {subItem.title && !subItem.icon && (
+                            <Text
+                              textStyle={subItem.textStyle}
+                              fontWeight={
+                                subItem.textStyle !== "body-1" ? 400 : undefined
+                              }
+                              color="chakra-body-text"
+                            >
+                              {subItem.title}
+                            </Text>
+                          )}
+                          {subItem.icon && (
+                            <MenuItem
+                              icon={subItem.icon}
+                              title={subItem.title}
+                              action={subItem.action}
+                              isActive={subItem.isActive}
+                            />
+                          )}
+                        </Menu.Item>
+                      ))}
+                    </Menu.List>
+                  </>
+                )
+              }}
+            </Menu>
+          )}
+
+          {item.type === "horizontal-list" && (
```

Comment:
nit: extract out a type, unless tiptap already provides a d.ts which makes this typesafe

Reply:
Hmm the type is defined by us, Tiptap isn't involved here. The type is defined earlier in the file, which also allows us to benefit from discriminated unions, thereby preserving the type-safety.